File: IOS/aqua/DiseaseClassifier.swift
//
//  DiseaseClassifier.swift
//  aqua
//
//  Created by aryaman jaiswal on 31/10/25.
//

import Foundation
import UIKit
import Combine

class DiseaseClassifier {
    static let shared = DiseaseClassifier()

    // Backend API configuration
    private let baseURL = "http://192.168.1.21:8000/api/v1"  // Change to your backend URL

    // Disease information database - mapped to the 7 outputs from fish_disease.keras model
    private let diseaseInfo: [String: (description: String, recommendations: [String])] = [
        "Bacterial Red disease": (
            description: "Bacterial hemorrhagic septicemia causing red lesions and hemorrhaging. Common bacterial infection in fish.",
            recommendations: [
                "Quarantine infected fish immediately",
                "Use antibiotics (oxytetracycline or florfenicol) as prescribed",
                "Improve water quality and increase aeration",
                "Consider salt bath treatment (1-3%)",
                "Consult with aquaculture veterinarian"
            ]
        ),
        "Bacterial diseases - Aeromoniasis": (
            description: "Systemic bacterial infection caused by Aeromonas species, leading to septicemia and organ damage.",
            recommendations: [
                "Antibiotic treatment (oxytetracycline, sulfonamides)",
                "Improve water quality immediately",
                "Increase dissolved oxygen levels",
                "Reduce stocking density if overcrowded",
                "Implement salt treatment protocol"
            ]
        ),
        "Bacterial gill disease": (
            description: "Bacterial infection affecting gill tissue, impairing respiratory function and leading to oxygen deprivation.",
            recommendations: [
                "Improve water quality immediately",
                "Increase aeration to boost dissolved oxygen",
                "Reduce feeding temporarily",
                "Apply antibiotics (chloramine-T, hydrogen peroxide bath)",
                "Monitor ammonia levels closely"
            ]
        ),
        "Fungal diseases Saprolegniasis": (
            description: "Fungal infection caused by Saprolegnia species, appearing as cotton-like growth on fish body, fins, or eggs.",
            recommendations: [
                "Salt bath treatment (0.5-1% for 10-15 minutes)",
                "Use antifungal medication (malachite green, methylene blue)",
                "Improve water quality and circulation",
                "Remove dead tissue if severe",
                "Handle fish carefully to avoid injuries"
            ]
        ),
        "Healthy Fish": (
            description: "Fish appears healthy with no visible signs of disease or distress. Continue regular monitoring and maintenance.",
            recommendations: [
                "Continue regular water quality monitoring",
                "Maintain optimal water parameters",
                "Follow regular feeding schedule",
                "Monitor weekly for any changes",
                "Keep up preventive care practices"
            ]
        ),
        "Parasitic diseases": (
            description: "External or internal parasites affecting fish health, including protozoa, worms, and crustaceans.",
            recommendations: [
                "Identify specific parasite type",
                "Treat with appropriate antiparasitic medication",
                "For Ich: raise temperature to 30°C with salt treatment",
                "Quarantine new fish for 2-3 weeks",
                "Disinfect equipment between uses"
            ]
        ),
        "Viral diseases White tail disease": (
            description: "Viral infection causing white discoloration of the tail and systemic illness. Highly contagious and often fatal.",
            recommendations: [
                "Quarantine infected fish immediately",
                "No specific antiviral treatment - supportive care only",
                "Cull severely affected fish to prevent spread",
                "Disinfect all equipment and tanks thoroughly",
                "Source fish from certified disease-free suppliers"
            ]
        )
    ]

    private init() {
        // No model loading needed - using backend API
    }
    
    func classifyImage(_ image: UIImage, completion: @escaping (Result<DiseaseResult, Error>) -> Void) {
        // Convert image to JPEG data
        guard let imageData = image.jpegData(compressionQuality: 0.8) else {
            completion(.failure(ClassificationError.invalidImage))
            return
        }

        // Create multipart form data request
        let url = URL(string: "\(baseURL)/analysis/disease-detection/ml-only")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"

        let boundary = UUID().uuidString
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()

        // Add image file
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"fish.jpg\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: image/jpeg\r\n\r\n".data(using: .utf8)!)
        body.append(imageData)
        body.append("\r\n".data(using: .utf8)!)
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)

        request.httpBody = body

        // Make the request
        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            if let error = error {
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
                return
            }

            guard let data = data else {
                DispatchQueue.main.async {
                    completion(.failure(ClassificationError.noResults))
                }
                return
            }

            do {
                // Parse JSON response
                let decoder = JSONDecoder()
                let apiResponse = try decoder.decode(DiseaseDetectionResponse.self, from: data)

                // Get the top prediction
                if let topPrediction = apiResponse.predictions.first {
                    // Debug logging
                    print("📱 iOS: Received prediction from backend:")
                    print("   Name: \(topPrediction.name)")
                    print("   Confidence: \(topPrediction.confidence)")
                    print("   Description from backend: \(topPrediction.description ?? "nil")")
                    print("   Prevention count: \(topPrediction.prevention?.count ?? 0)")

                    let diseaseResult = self?.createDiseaseResult(from: topPrediction) ?? DiseaseResult(
                        diseaseName: "Unknown",
                        confidence: 0.0,
                        description: "Unable to classify the image",
                        recommendations: ["Please try with a clearer image", "Consult with a fish health expert"],
                        recommendedProducts: []
                    )

                    DispatchQueue.main.async {
                        completion(.success(diseaseResult))
                    }
                } else {
                    DispatchQueue.main.async {
                        completion(.failure(ClassificationError.noResults))
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }.resume()
    }
    
    private func createDiseaseResult(from prediction: DiseasePrediction) -> DiseaseResult {
        // Use backend data if available, otherwise fall back to local data
        let localInfo = diseaseInfo[prediction.name]

        let description: String
        let recommendations: [String]

        // Prefer backend description if available
        if let backendDescription = prediction.description, !backendDescription.isEmpty {
            description = backendDescription
            print("   ✓ Using description from backend")
        } else if let localDesc = localInfo?.description {
            description = localDesc
            print("   ⚠️ Using LOCAL description (backend sent none)")
        } else {
            description = "Disease detected but specific information not available."
            print("   ⚠️ Using FALLBACK description")
        }

        // Prefer backend prevention/recommendations if available
        if let backendPrevention = prediction.prevention, !backendPrevention.isEmpty {
            recommendations = backendPrevention
            print("   ✓ Using \(backendPrevention.count) recommendations from backend")
        } else if let localRecs = localInfo?.recommendations {
            recommendations = localRecs
            print("   ⚠️ Using LOCAL recommendations (backend sent none)")
        } else {
            recommendations = ["Consult with a fish health specialist", "Monitor fish closely", "Improve water quality"]
            print("   ⚠️ Using FALLBACK recommendations")
        }

        return DiseaseResult(
            diseaseName: prediction.name,
            confidence: prediction.confidence,
            description: description,
            recommendations: recommendations,
            recommendedProducts: getRecommendedProducts(for: prediction.name)
        )
    }
    
    private func getRecommendedProducts(for diseaseName: String) -> [MarketplaceProduct] {
        // Get products from marketplace based on disease type
        let allProducts = MarketplaceProduct.sampleProducts

        switch diseaseName {
        case "Healthy Fish":
            return allProducts.filter { product in
                product.category == .feed ||
                (product.category == .maintenance && product.name.lowercased().contains("water"))
            }.prefix(3).map { $0 }

        case "Bacterial Red disease", "Bacterial diseases - Aeromoniasis", "Bacterial gill disease":
            return allProducts.filter { product in
                product.category == .ointments &&
                (product.name.lowercased().contains("antibacterial") ||
                 product.name.lowercased().contains("antibiotic") ||
                 product.description.lowercased().contains("bacterial"))
            }.prefix(4).map { $0 }

        case "Fungal diseases Saprolegniasis":
            return allProducts.filter { product in
                product.category == .ointments &&
                (product.name.lowercased().contains("antifungal") ||
                 product.name.lowercased().contains("fungus") ||
                 product.description.lowercased().contains("fungal"))
            }.prefix(4).map { $0 }

        case "Parasitic diseases":
            return allProducts.filter { product in
                product.category == .ointments &&
                (product.name.lowercased().contains("antiparasitic") ||
                 product.name.lowercased().contains("parasite") ||
                 product.description.lowercased().contains("parasitic"))
            }.prefix(4).map { $0 }

        case "Viral diseases White tail disease":
            return allProducts.filter { product in
                (product.category == .feed && product.description.lowercased().contains("immunity")) ||
                (product.category == .ointments && product.description.lowercased().contains("viral"))
            }.prefix(3).map { $0 }

        default:
            // For unknown diseases, return general health products
            return allProducts.filter { product in
                product.category == .ointments || product.category == .feed
            }.prefix(3).map { $0 }
        }
    }
}

// MARK: - API Response Models

struct DiseaseDetectionResponse: Codable, Sendable {
    let predictions: [DiseasePrediction]
    let message: String?
}

struct DiseasePrediction: Codable, Sendable {
    let name: String
    let confidence: Double
    let description: String?
    let causes: [String]?
    let symptoms: [String]?
    let treatment: String?
    let prevention: [String]?
}

// MARK: - Error Types

enum ClassificationError: LocalizedError {
    case modelNotLoaded
    case invalidImage
    case noResults

    var errorDescription: String? {
        switch self {
        case .modelNotLoaded:
            return "Unable to connect to the disease detection service. Please check your internet connection."
        case .invalidImage:
            return "The selected image could not be processed. Please try selecting a different image."
        case .noResults:
            return "No classification results were returned. Please try with a clearer image of the fish."
        }
    }
}

File: aquasense_backend/ml/disease_classifier.py
"""
Disease Classifier - ML model for fish disease detection
"""
import os
import base64
import io
import json
import logging
from typing import List, Optional
from PIL import Image
import numpy as np

from config.settings import settings
from ml.preprocessing import preprocess_image
from schemas.analysis import DiseaseInfo

logger = logging.getLogger(__name__)


class DiseaseClassifier:
    """
    Fish disease classifier using TensorFlow/Keras model.
    Model is provided by the ML team.
    """

    def __init__(self):
        """Initialize the disease classifier"""
        self.model = None
        self.model_path = settings.DISEASE_MODEL_PATH
        self.label_map_path = os.path.join(os.path.dirname(self.model_path), "label_map.json")
        self.label_map = self._load_label_map()
        self.disease_info = self._load_disease_info()

        # Load model if it exists
        if os.path.exists(self.model_path):
            self._load_model()
        else:
            logger.warning(f"Disease model not found at {self.model_path}. Model inference will be skipped.")

    def _load_model(self):
        """Load the Keras model"""
        try:
            import tensorflow as tf
            self.model = tf.keras.models.load_model(self.model_path)
            logger.info("Disease classification model loaded successfully")
        except Exception as e:
            logger.error(f"Failed to load model: {str(e)}")
            self.model = None

    def _load_label_map(self) -> dict:
        """
        Load label map from JSON file.
        Maps class indices to disease names.
        """
        if os.path.exists(self.label_map_path):
            try:
                with open(self.label_map_path, 'r') as f:
                    label_map = json.load(f)
                logger.info(f"Label map loaded successfully with {len(label_map)} classes")
                return label_map
            except Exception as e:
                logger.error(f"Error loading label map: {str(e)}")
                return self._get_default_label_map()
        else:
            logger.warning(f"Label map not found at {self.label_map_path}. Using default mapping.")
            return self._get_default_label_map()

    def _get_default_label_map(self) -> dict:
        """
        Get default label map if file not found.
        """
        return {
            "0": "Bacterial Red disease",
            "1": "Bacterial diseases - Aeromoniasis",
            "2": "Bacterial gill disease",
            "3": "Fungal diseases Saprolegniasis",
            "4": "Healthy Fish",
            "5": "Parasitic diseases",
            "6": "Viral diseases White tail disease"
        }

    def _load_disease_info(self) -> dict:
        """
        Load disease information mapping.
        Maps disease names from label_map to detailed disease information.
        """
        return {
            "Bacterial Red disease": {
                "name": "Bacterial Red Disease",
                "description": "Bacterial hemorrhagic septicemia causing red lesions and hemorrhaging. Common bacterial infection in fish.",
                "causes": ["Aeromonas bacteria", "Poor water quality", "Stress", "Physical injuries", "Weakened immune system"],
                "symptoms": ["Red lesions on body", "Hemorrhages on skin and fins", "Ulcers", "Lethargy", "Loss of appetite", "Abnormal swimming"],
                "treatment": "Quarantine infected fish immediately. Use antibiotics (oxytetracycline or florfenicol) as prescribed. Improve water quality. Increase aeration. Salt bath (1-3%).",
                "prevention": ["Maintain excellent water quality", "Regular water changes", "Avoid overcrowding", "Quarantine new fish", "Reduce stress factors", "Proper nutrition"]
            },
            "Bacterial diseases - Aeromoniasis": {
                "name": "Aeromoniasis (Motile Aeromonas Septicemia)",
                "description": "Systemic bacterial infection caused by Aeromonas species, leading to septicemia and organ damage.",
                "causes": ["Aeromonas hydrophila bacteria", "Poor water quality", "Stress", "High organic load", "Temperature fluctuations"],
                "symptoms": ["Hemorrhages on body", "Fin rot", "Ulcers", "Swollen abdomen", "Exophthalmia (bulging eyes)", "Lethargy", "Loss of appetite"],
                "treatment": "Antibiotic treatment (oxytetracycline, sulfonamides). Improve water quality immediately. Increase dissolved oxygen. Reduce stocking density. Salt treatment.",
                "prevention": ["Maintain optimal water parameters", "Regular monitoring", "Avoid stress", "Proper feeding", "Quarantine protocols", "Disinfect equipment"]
            },
            "Bacterial gill disease": {
                "name": "Bacterial Gill Disease",
                "description": "Bacterial infection affecting gill tissue, impairing respiratory function and leading to oxygen deprivation.",
                "causes": ["Flavobacterium branchiophilum", "Poor water quality", "High ammonia levels", "Overcrowding", "Low dissolved oxygen"],
                "symptoms": ["Rapid gill movement", "Gasping at surface", "Pale or swollen gills", "Excess mucus on gills", "Lethargy", "Reduced feeding", "Mortality"],
                "treatment": "Improve water quality immediately. Increase aeration. Reduce feeding temporarily. Antibiotics (chloramine-T, hydrogen peroxide bath). Copper sulfate treatment (carefully dosed).",
                "prevention": ["Maintain high dissolved oxygen (>5 mg/L)", "Keep ammonia at 0 ppm", "Regular water changes", "Avoid overcrowding", "Proper filtration", "Monitor water parameters"]
            },
            "Fungal diseases Saprolegniasis": {
                "name": "Saprolegniasis (Fungal Infection)",
                "description": "Fungal infection caused by Saprolegnia species, appearing as cotton-like growth on fish body, fins, or eggs.",
                "causes": ["Saprolegnia fungus (opportunistic)", "Physical injury", "Poor water quality", "Low temperature", "Weakened immune system", "Stress"],
                "symptoms": ["White or gray cotton-like growth", "Fluffy patches on body/fins", "Fin deterioration", "Often follows injury or stress", "Secondary to bacterial infections"],
                "treatment": "Salt bath (0.5-1% for 10-15 minutes). Antifungal medication (malachite green, methylene blue, potassium permanganate). Improve water quality. Remove dead tissue if severe.",
                "prevention": ["Avoid physical injuries", "Maintain good water quality", "Proper nutrition", "Handle fish carefully", "Quarantine injured fish", "Maintain optimal temperature"]
            },
            "Healthy Fish": {
                "name": "Healthy Fish - No Disease Detected",
                "description": "Fish appears healthy with no visible signs of disease or distress. Continue regular monitoring and maintenance.",
                "causes": [],
                "symptoms": ["Active swimming", "Normal appetite", "Bright coloration", "Clear eyes", "Intact fins", "Normal gill movement"],
                "treatment": "No treatment needed. Continue regular monitoring and maintenance practices.",
                "prevention": ["Maintain optimal water quality", "Regular feeding schedule", "Avoid overcrowding", "Monitor water parameters weekly", "Quarantine new fish", "Reduce stress factors"]
            },
            "Parasitic diseases": {
                "name": "Parasitic Infections",
                "description": "External or internal parasites affecting fish health, including protozoa, worms, and crustaceans.",
                "causes": ["Ichthyophthirius (Ich)", "Trichodina", "Gyrodactylus", "Dactylogyrus", "Anchor worms", "Fish lice", "Introduction of infected fish", "Poor quarantine"],
                "symptoms": ["White spots on body (Ich)", "Scratching/flashing behavior", "Excess mucus production", "Rapid gill movement", "Weight loss", "Lethargy", "Clamped fins"],
                "treatment": "Identify specific parasite. For Ich: raise temperature to 30°C, salt treatment (1-2%), copper sulfate. For external parasites: formalin bath, potassium permanganate. For internal parasites: medicated feed.",
                "prevention": ["Quarantine new fish for 2-3 weeks", "Disinfect equipment", "Regular inspection", "Maintain water quality", "Avoid stress", "Clean equipment between uses"]
            },
            "Viral diseases White tail disease": {
                "name": "Viral White Tail Disease",
                "description": "Viral infection causing white discoloration of the tail and systemic illness. Highly contagious and often fatal.",
                "causes": ["Viral pathogen (species-specific virus)", "Infected fish introduction", "Contaminated water", "Stress", "Poor biosecurity"],
                "symptoms": ["White discoloration of tail", "Tail necrosis", "Lethargy", "Loss of appetite", "Abnormal swimming", "High mortality rate", "Systemic infection"],
                "treatment": "No specific antiviral treatment. Supportive care only. Quarantine infected fish immediately. Cull severely affected fish to prevent spread. Disinfect all equipment and tanks.",
                "prevention": ["Source fish from certified disease-free suppliers", "Strict quarantine protocols (3-4 weeks)", "Excellent biosecurity", "Do not share equipment between tanks", "Maintain optimal conditions to boost immunity", "Vaccination if available"]
            }
        }

    async def predict(self, image_base64: str) -> List[DiseaseInfo]:
        """
        Predict disease from fish image.

        Args:
            image_base64: Base64 encoded image

        Returns:
            List of DiseaseInfo objects with predictions
        """
        if not self.model:
            logger.warning("Model not loaded. Returning empty predictions.")
            return []

        try:
            # Decode and preprocess image
            image_bytes = base64.b64decode(image_base64)
            image = Image.open(io.BytesIO(image_bytes))
            processed_image = preprocess_image(image)

            # Make prediction
            predictions = self.model.predict(processed_image)

            # Convert predictions to DiseaseInfo objects
            results = self._predictions_to_disease_info(predictions[0])

            return results

        except Exception as e:
            logger.error(f"Error in disease prediction: {str(e)}")
            return []

    def _predictions_to_disease_info(
        self,
        predictions: np.ndarray,
        confidence_threshold: float = 0.2
    ) -> List[DiseaseInfo]:
        """
        Convert model predictions to DiseaseInfo objects.

        Args:
            predictions: Model output probabilities
            confidence_threshold: Minimum confidence to include in results

        Returns:
            List of DiseaseInfo objects
        """
        results = []

        # Get top predictions
        sorted_indices = np.argsort(predictions)[::-1]

        # Log raw ML predictions before filtering
        print("\nRAW ML MODEL OUTPUT (sorted by confidence):")
        for i, idx in enumerate(sorted_indices[:5], 1):  # Show top 5
            disease_name = self.label_map.get(str(idx), "Unknown")
            confidence = float(predictions[idx])
            print(f"  {i}. [{idx}] {disease_name}: {confidence:.2%}")
        print()

        for idx in sorted_indices:
            confidence = float(predictions[idx])

            # Skip if below threshold
            if confidence < confidence_threshold:
                print(f"  ⊘ Skipped (below threshold {confidence_threshold}): confidence {confidence:.2%}")
                continue

            # Get disease name from label map
            disease_name = self.label_map.get(str(idx))

            if not disease_name:
                logger.warning(f"No label found for index {idx}")
                continue

            # Skip "Healthy Fish" unless it has very high confidence
            if disease_name == "Healthy Fish" and confidence < 0.7:
                print(f"  ⊘ Skipped '{disease_name}': {confidence:.2%} (< 70% threshold for healthy)")
                continue

            # Get detailed disease information
            disease_data = self.disease_info.get(disease_name, {})

            if not disease_data:
                # Fallback if no detailed info available
                disease_data = {
                    "name": disease_name,
                    "description": f"Detected {disease_name}",
                    "causes": [],
                    "symptoms": [],
                    "treatment": "Consult aquaculture veterinarian for specific treatment.",
                    "prevention": ["Maintain good water quality", "Regular monitoring"]
                }

            disease_info = DiseaseInfo(
                name=disease_data.get("name", disease_name),
                confidence=confidence,
                description=disease_data.get("description", ""),
                causes=disease_data.get("causes", []),
                symptoms=disease_data.get("symptoms", []),
                treatment=disease_data.get("treatment", ""),
                prevention=disease_data.get("prevention", [])
            )

            results.append(disease_info)
            print(f"  ✓ Added to results: {disease_name} ({confidence:.2%})")

            # Limit to top 3 results
            if len(results) >= 3:
                break

        print(f"\nFINAL FILTERED RESULTS: {len(results)} diseases returned")
        if results:
            print(f"FIRST RESULT (sent to iOS app): {results[0].name} ({results[0].confidence:.2%})")
        print()

        return results

    def get_disease_info(self, disease_name: str) -> Optional[dict]:
        """
        Get detailed information about a specific disease.

        Args:
            disease_name: Name of the disease

        Returns:
            Disease information dictionary or None
        """
        return self.disease_info.get(disease_name)
